Add error end functionality

AntonVucinic opened the pull request:
Code change for issue #61.

I also noticed that the field `code` of  [`Diagnositc`](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#diagnostic) is unused but rather errors were parsed in such a way that the mypy error code was included in the diagnostic message. I think this is a bug and am also willing to fix it.

AntonVucinic commented:
According to [LSP docs](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#range) the end position is exclusive so I just fixed that with the second commit

Richardk2n commented on pylsp_mypy/plugin.py:
```diff
@@ -90,25 +90,26 @@ def parse_line(line: str, document: Optional[Document] = None) -> Optional[Dict[
+        if diag["range"]["end"]["character"] is None:
+            diag["range"]["end"]["character"] = (
+                offset + len(word)
+                if (document and (word := document.word_at_position(diag["range"]["start"])))
```
Walrus operator was introduced with 3.8, we still support 3.7

rchl commented on pylsp_mypy/plugin.py:
```diff
@@ -90,25 +90,24 @@ def parse_line(line: str, document: Optional[Document] = None) -> Optional[Dict[
                 "start": {"line": lineno, "character": offset},
-                # There may be a better solution, but mypy does not provide end
-                "end": {"line": lineno, "character": offset + 1},
+                "end": {"line": end_lineno or lineno, "character": end_offset},
```
Those can be `None` and `None` is not allowed in `Position`.
I see that you later adjust `character` if it's `None` but not `line`.
You should probably fall back to previous behavior in that case.

Richardk2n replied:
So mypy with this option behaves differently in 3.7? (Tests seem to indicate that, I just want to confirm)
Is this documented in mypy as intended or is this a bug?

Either way we will not write a lengthy workaround as 3.7 support will likely be dropped soon anyway.

AntonVucinic replied:
Didn't find anything in docs or an issue but I did try with multiple different error types and they all had the same behaviour.

Can you clarify what you mean by not supporting a lengthy workaround.

Richardk2n replied:
Maybe we should ask the mypy devs if this is intentional or a bug.

Mypy seems to behave differently in python 3.7 than in the other versions. In order to handle this, 'unreadable code' (as you put it) is required. This is what I mean by 'lengthy workaround'. Having this in the codebase is not desirable in the first place, but in this case there is another point to consider. Python 3.7 will be end of life very soon. Mypy will most likely drop support for it then and so will we. This means, that in approx two months our code does not need to work with python 3.7 anymore. So there is little reason to try really hard to handle some special case in 3.7 now.

So: If the test for 3.7 does not work, just ignore it for the time being.

Cleaning up the regex is fine with me and probably a good idea.

rchl commented on pylsp_mypy/plugin.py:
```diff
@@ -90,25 +90,24 @@ def parse_line(line: str, document: Optional[Document] = None) -> Optional[Dict[
-            word = document.word_at_position(diag["range"]["start"])
-            if word:
-                diag["range"]["end"]["character"] = diag["range"]["start"]["character"] + len(word)
+        if diag["range"]["end"]["character"] is None:
+            if document:
+                word = document.word_at_position(diag["range"]["start"])
+                diag["range"]["end"]["character"] = offset + len(word) if word else offset + 1
```
Maybe move that code to after where `end_offset` is defined instead of adjusting diagnostic after creating it?

AntonVucinic commented:
Sorry for the wait but I fixed it in what I think is the best way possible and I also manually ran tests in all 5 versions and all seems to work fine. It is a lot of changes though.

In short I removed the redundant tests that were testing for lines with partial information. I also modified the other tests so that they check that other fields of diagnostic are also correctly parsed.

Richardk2n commented:
Don't worry. I appreciate every contribution.

How did you handle the python 3.7 I don't see any special code for that?

AntonVucinic commented:
As you said I didn't. I only modified the test case that depended on the actual output of mypy so taht it would pass in 3.7. The other tests use predefined output lines so they were fine.

Richardk2n commented:
Fyi, I made a bug report at https://github.com/python/mypy/issues/15281 to assure 3.7 is actually broken, and we did not misunderstand something.
Other than that I am rather happy with your changes and will merge them soon. Sorry for the delay.
